snoop/data/digests.py

```diff
@@ -117,17 +118,24 @@ def gather(blob, **depends_on):
                 rv['ocrimage'] = True
         rv['ocrtext'] = ocr_results
 
-    # combine texts and detect language
-    if settings.DETECT_LANGUAGE:
+    #TODO: Add entity Extraction setting to snoop settings
+    if settings.EXTRACT_ENTITIES:
+        text = rv.get('text', '')
+        ocrtexts = [v for v in rv.get('ocrtext', {}).values() if v]
+        if text or ocrtexts:
+            alltext = text + "\n" + "\n".join(ocrtexts)
+            nlp_response = entities.get_entities(alltext)
```

Comment:
Get the entities separately for each of the `text` and `ocrtext.*` because they are shown separately in the UI tab.


All your responses will also appear in the index (here, the `rv`) in fields keyed under the text identifier.

All the data you get from your service should be saved here into the Entity table before going forward.

Reply:
Okay. So:

- [ ] 1.  One Request per text (text and OCR text)
- [ ] 2.  Save all Entities in Entity table (unique Combination of label and text)
- [ ] 3. Store the start and end in EntityHit Table
One question about storing the stuff for EntityHit. What should I put in rv for it? Just all entities with all info?

Reply:
The keys `rv` go down to be served through the API and be stored to elasticsearch. This doesn't happen directly, there are some more functions that pick given fields to serve in the API or store in the index.

Here we always want to have the same data and structure, so let's think of some stuff we want in the index first for every document:

- `rv['entity']`: list of all the entities (de-duplicated) of all types (the interesting types, that we want in a single bucket - not dates)
- `rv['extracted-date']`: list of dates extracted from the document
- `rv['extracted-geolocation']`: list of GPS positions data extracted from the document (https://www.elastic.co/guide/en/elasticsearch/reference/6.8/geo-point.html to be used later with https://www.elastic.co/guide/en/elasticsearch/reference/6.8/geo-queries.html) - we need a local database of location names vs. GPS, so this may be out of scope for your project
- `rv['ent-type.X']`: for each entity type, a de-duplicated list of the texts
- `rv['ent-id.X']`: for each entity hit, put the numerical ID here


And in snoop also add these fields to the [mapping](https://github.com/liquidinvestigations/hoover-snoop2/blob/8a6dbdf1fd1ea56db386628f0559b097cdaa1a61/snoop/data/indexing.py#L22):
- `entity`: text (we'll use the `.keyword` generated sub-field for buckets, and use this for fuzzy searching done by the user, like `entity:Jim~2`)
- `extracted-date`: date
- `extracted-geolocation`: geo-point
- `ent-type`: object
- `ent-type.*`: text (see tags field for dynamic mapping, but use text instead of keyword)
- `ent-id.*`: int (same as above)

Be sure to forward these fields in [the ugly transcription function here](https://github.com/liquidinvestigations/hoover-snoop2/blob/8a6dbdf1fd1ea56db386628f0559b097cdaa1a61/snoop/data/digests.py#L489) that converts your `rv` (here called `digest_data`) and the state of the File objects into the stuff that gets indexed by es and presented on the API.